frontends/halo2/src/backend/llzk/factory.rs

```diff
@@ -1,96 +1,87 @@
-            iter::repeat($base)
-                .map(Field::from)
-                .zip(0..self.$name)
-                .map(|(f, n)| f.renamed(|name| format!("{name}_{n}")))
-        }
-    };
-}
-
 impl StructIO {
-    fn fields(&self) -> impl IntoIterator<Item = Field> {
-        self.private_inputs()
-            .chain(self.private_outputs())
-            .chain(self.public_outputs())
+    fn fields<'c>(
+        &self,
+        context: &'c Context,
+        header: &str,
+    ) -> impl Iterator<Item = Result<FieldDefOp<'c>, LlzkError>> {
+        let public_filename = format!("struct {header} | public outputs");
+        let private_filename = format!("struct {header} | private outputs");
+        std::iter::repeat_n(true, self.public_outputs)
+            .enumerate()
+            .chain(std::iter::repeat_n(false, self.private_outputs).enumerate())
+            .map(move |(n, public)| {
+                let filename = if public {
+                    &public_filename
+                } else {
+                    &private_filename
+                };
+                (public, Location::new(context, filename, n, 0))
+            })
+            .enumerate()
+            .map(|(n, (public, loc))| {
+                let name = format!("out_{n}");
+                r#struct::field(loc, &name, FeltType::new(context), false, public)
+            })
     }
 
-    field_iter!(private_inputs, ("in", false));
-    field_iter!(private_outputs, ("out", false));
-    field_iter!(public_outputs, ("out", true));
+    pub fn args<'c>(
+        &self,
+        ctx: &'c Context,
+        is_main: bool,
+        struct_name: &str,
+    ) -> Vec<(Type<'c>, Location<'c>)> {
+        let public_filename = format!("struct {struct_name} | public inputs");
+        let private_filename = format!("struct {struct_name} | private inputs");
+        let public_locs = std::iter::repeat(&public_filename)
+            .enumerate()
+            .take(self.public_inputs);
+        let private_locs = std::iter::repeat(&private_filename)
+            .enumerate()
+            .take(self.private_inputs);
+        let locs = public_locs
+            .chain(private_locs)
+            .map(|(n, filename)| Location::new(ctx, filename, n, 0));
+
+        let ty: Type<'c> = if is_main {
+            StructType::from_str(ctx, "Signal").into()
+        } else {
+            FeltType::new(ctx).into()
+        };
+        let types = std::iter::repeat(ty).take(self.public_inputs + self.private_inputs);
+
+        std::iter::zip(types, locs).collect()
+    }
 
-    pub fn public_inputs<'c>(&self, ctx: &'c Context) -> impl IntoIterator<Item = Type<'c>> {
-        iter::repeat_with(|| FeltType::new(ctx).into()).take(self.public_inputs)
+    /// Returns the list of argument attributes for the struct's functions.
+    pub fn arg_attrs<'c>(&self, ctx: &'c Context) -> Vec<Vec<NamedAttribute<'c>>> {
+        let pub_attr = (
+            Identifier::new(ctx, "llzk.pub"),
```

Comment:
Is there a way to get the attribute name from the llzk-lib itself instead of baking in the string? I realize this may be annoying with the C API, so no worries if doing it that way would be a big deal.